File: IDATT2106 Systemutvikling 2 med smidig prosjekt/backend/src/main/java/com/ntnu/gidd/util/JwtUtil.java
package com.ntnu.gidd.util;

import com.ntnu.gidd.security.config.JWTConfig;
import com.ntnu.gidd.security.token.JwtRefreshToken;
import com.ntnu.gidd.security.token.RawJwtAccessToken;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.MalformedJwtException;
import io.jsonwebtoken.SignatureException;
import io.jsonwebtoken.UnsupportedJwtException;
import lombok.AllArgsConstructor;
import org.springframework.security.authentication.BadCredentialsException;

import java.util.Optional;
import java.util.UUID;


@AllArgsConstructor
public class JwtUtil {
	
	private JWTConfig jwtConfig;
	
	public String getEmailFromToken(String token) {
		return Jwts.parser()
				.setSigningKey(jwtConfig.getSecret())
				.parseClaimsJws(token.replace(jwtConfig.getPrefix(), ""))
				.getBody().getSubject();
	}

	public boolean isValidToken(String token) {
		try {
			Jwts.parser()
					.setSigningKey(jwtConfig.getSecret())
					.parseClaimsJws(token.replace(jwtConfig.getPrefix(), ""));
			return true;
		} catch (SignatureException | MalformedJwtException | UnsupportedJwtException | IllegalArgumentException ex) {
			throw new BadCredentialsException("Invalid credentials", ex);
		}
	}

	public Optional<JwtRefreshToken> parseToken(String token) {
		RawJwtAccessToken rawJwtAccessToken = new RawJwtAccessToken(token);
		return JwtRefreshToken.of(rawJwtAccessToken, jwtConfig.getSecret());
	}

	public UUID getUserIdFromToken(String token) {
		Object uuid = Jwts.parser()
				.setSigningKey(jwtConfig.getSecret())
				.parseClaimsJws(token.replace(jwtConfig.getPrefix(), ""))
				.getBody()
				.get("uuid");
		return UUID.fromString(String.valueOf(uuid));
	}
}

File: IDATT2105 Full-stack applikasjonsutvikling/Rombestilling/backend/src/main/kotlin/ntnu/idatt2105/security/JwtUtil.kt
package ntnu.idatt2105.security

import io.jsonwebtoken.Jwts
import io.jsonwebtoken.MalformedJwtException
import io.jsonwebtoken.UnsupportedJwtException
import io.jsonwebtoken.security.SignatureException
import ntnu.idatt2105.core.util.SecurityConstants.AUTHORITIES_KEY
import ntnu.idatt2105.security.config.JWTConfig
import ntnu.idatt2105.security.token.JwtRefreshToken
import ntnu.idatt2105.security.token.RawJwtAccessToken
import org.springframework.security.authentication.BadCredentialsException
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.stereotype.Component
import java.util.*

@Component
class JwtUtil(val jwtConfig: JWTConfig) {

    fun getEmailFromToken(token: String): String? {
        return Jwts.parserBuilder()
                .setSigningKey(jwtConfig.secret).build()
                .parseClaimsJws(token.replace(jwtConfig.prefix, ""))
                .body.subject
    }

    fun isValidToken(token: String): Boolean {
        return try {
            Jwts.parserBuilder()
                    .setSigningKey(jwtConfig.secret).build()
                    .parseClaimsJws(token.replace(jwtConfig.prefix, ""))
            true
        } catch (ex: SignatureException) {
            throw BadCredentialsException("Invalid credentials", ex)
        } catch (ex: MalformedJwtException) {
            throw BadCredentialsException("Invalid credentials", ex)
        } catch (ex: UnsupportedJwtException) {
            throw BadCredentialsException("Invalid credentials", ex)
        } catch (ex: IllegalArgumentException) {
            throw BadCredentialsException("Invalid credentials", ex)
        }
    }

    fun parseToken(token: String?): JwtRefreshToken? {
        val rawJwtAccessToken = RawJwtAccessToken(token!!)
        return JwtRefreshToken.of(rawJwtAccessToken, jwtConfig.secret)
    }

    fun getUserIdFromToken(token: String): UUID? {
        val uuid = Jwts.parserBuilder()
                .setSigningKey(jwtConfig.secret).build()
                .parseClaimsJws(token.replace(jwtConfig.prefix, ""))
                .body["uuid"]
        return UUID.fromString(uuid.toString())
    }

    fun getRolesFromToken(token: String) =
         Jwts.parserBuilder()
            .setSigningKey(jwtConfig.secret).build()
            .parseClaimsJws(token.replace(jwtConfig.prefix, ""))
            .body[AUTHORITIES_KEY].toString().split(",")
            .map { SimpleGrantedAuthority(it) }
}
